crypto/armcap.c

```diff
@@ -168,42 +174,58 @@ void OPENSSL_cpuid_setup(void)
             if (hwcap & HWCAP_CE_SHA256)
                 OPENSSL_armcap_P |= ARMV8_SHA256;
 
-# ifdef __aarch64__
+#  ifdef __aarch64__
             if (hwcap & HWCAP_CE_SHA512)
                 OPENSSL_armcap_P |= ARMV8_SHA512;
-# endif
+#  endif
         }
     } else if (sigsetjmp(ill_jmp, 1) == 0) {
-        _armv7_neon_probe();
-        OPENSSL_armcap_P |= ARMV7_NEON;
-        if (sigsetjmp(ill_jmp, 1) == 0) {
-            _armv8_pmull_probe();
-            OPENSSL_armcap_P |= ARMV8_PMULL | ARMV8_AES;
-        } else if (sigsetjmp(ill_jmp, 1) == 0) {
-            _armv8_aes_probe();
-            OPENSSL_armcap_P |= ARMV8_AES;
-        }
-        if (sigsetjmp(ill_jmp, 1) == 0) {
-            _armv8_sha1_probe();
-            OPENSSL_armcap_P |= ARMV8_SHA1;
-        }
-        if (sigsetjmp(ill_jmp, 1) == 0) {
-            _armv8_sha256_probe();
-            OPENSSL_armcap_P |= ARMV8_SHA256;
-        }
-# if defined(__aarch64__) && !defined(__APPLE__)
-        if (sigsetjmp(ill_jmp, 1) == 0) {
-            _armv8_sha512_probe();
-            OPENSSL_armcap_P |= ARMV8_SHA512;
+        if (mask & ARMV7_NEON) {
+            _armv7_neon_probe();
+            OPENSSL_armcap_P |= ARMV7_NEON;
+            if (sigsetjmp(ill_jmp, 1) == 0) {
+                if (mask & ARMV8_PMULL) {
+                    _armv8_pmull_probe();
+                    OPENSSL_armcap_P |= ARMV8_PMULL | ARMV8_AES;
+                }
+            } else if (sigsetjmp(ill_jmp, 1) == 0) {
```

Comment:
Outcome would be wrong if you clear PMULL and set AES. Because it will never probe for AES. However, instead of treating it as special case, it might be appropriate to simply switch to `(mask & CAP) && sigsetjmp()` in all cases. Not only will it fix this corner case, it would be more readable. I mean sigsetjmp is effectively part of the probe, so you first decide if you should perform probe, then perform it. As opposite to making half way into probe then deciding if you should proceed to final part of the probe.

Reply:
I read the C90 standard very carefully, and it says setjmp __must__ be called only in one of these ways:

```
   An invocation of the setjmp macro shall appear only in one of the
following contexts:

 * the entire controlling expression of a selection or iteration statement;

 * one operand of a relational or equality operator with the other
   operand an integral constant expression, with the resulting expression
   being the entire controlling expression of a selection or iteration
   statement;

 * the operand of a unary ! operator with the resulting expression
   being the entire controlling expression of a selection or iteration
   statement; or

 * the entire expression of an expression statement (possibly cast to void).
```

so the suggested syntax is non-compliant.

Reply:
Would `if (mask & (ARMV8_PMULL|ARMV8_AES))` do the right thing?

Reply:
This should work:
```
if (mask & ...) 
   if (sigsetjmp(....) 
```

Reply:
Right, keyword is to flip conditions.